### PR TITLE
feat(mcp): group/public access control for MCP servers

### PR DESCRIPTION
**Pull request by beraterkanelcelik**
## Description

MCP servers currently have no way to restrict which users can add their tools to agents — every authenticated user sees every configured MCP server in the agent builder. This adds optional group/public access control.

It reuses the existing `MCPServer__User` / `MCPServer__UserGroup` association tables (they already existed in the schema; `get_mcp_servers_accessible_to_user` had a `TODO` and ignored them) and mirrors the existing `Persona` / `DocumentSet` / `LLMProvider` access pattern (`is_public` + groups via a versioned writer).

Behavior:
- Admins can mark an MCP server **public** (any user) or restrict it to specific **user groups** (and/or users) on the create/edit form.
- Only users with access see the server in the agent builder, and `upsert_persona` rejects attaching tools from a server the acting user can't access — so the gate can't be bypassed by calling the API directly.
- **Backward compatible:** existing servers, and any server with no restriction, default to `is_public = true`, so everyone keeps access.

### Layers (happy to split into stacked PRs if you'd rather)
1. migration + `is_public` model column
2. access read (`get_mcp_servers_accessible_to_user` / `user_can_access_mcp_server`) + API wiring + persona security gate
3. frontend `IsPublicGroupSelector` on the create/edit form

### EE note
The group/user **row writer** (`make_mcp_server_private`) is the only enterprise piece. Per the usual arrangement it's intentionally left out of this PR for the `ee/` side — CE ships the model, migration, read path, versioned hook, API, security gate, and UI. The CE stub raises `NotImplementedError` if a restriction is requested, exactly like `make_doc_set_private`. Happy to send the short spec for it.

## How Tested

- External-dependency unit tests (real Postgres) for the read path, the persona security gate, and edge cases (user in multiple groups, server restricted to multiple groups, admin-only private server, direct + group grants coexisting) — 12 tests, all pass under CE with no EE enabled.
- Playwright E2E (`web/tests/e2e/mcp/mcp_group_access.spec.ts`): restricted server persists `is_public=false` + groups, public default, editing restricted→public clears groups, and the Add-Server modal exposes the access selector.
- Manually verified end-to-end across member / non-member / admin users.

## Backporting (check the box to nominate this PR for backporting)

- [ ] This PR should be backported (currently 1.0.x)

---
## Summary by cubic
Adds public/group access control for MCP servers so admins can limit who can attach their tools. Users only see and use servers they can access; tool listings and persona updates enforce it, and CE responds with a clean 403 when restrictions are requested.

- **New Features**
  - Added `is_public` to MCPServer (default true); reused existing `MCPServer__User`/`MCPServer__UserGroup` link tables.
  - Filtered the user-facing server list to accessible servers (public, direct, or via group); admins see all, owners see their own.
  - Blocked persona updates from attaching tools from servers the user can’t access.
  - Gated non-admin tool listing (`GET /mcp/server/{id}/tools`) and filtered `GET /tools` to MCP tools from accessible servers.
  - API requests/responses include `is_public` and `groups`; the auth-config path leaves access unchanged unless explicitly set.
  - Create/edit form includes `IsPublicGroupSelector`; public servers send no groups.
  - CE returns 403 `EE_REQUIRED` on restrict; EE reconciles grants via versioned `make_mcp_server_private`.

- **Migration**
  - Run the DB migration adding `is_public` (existing servers remain public).
  - Implement the EE `make_mcp_server_private` hook on the `ee/` side if using groups.
  - No action needed for existing servers; visibility is unchanged by default.
  - Migration parent updated to the current Alembic head to avoid split heads.

<sup>Written for commit caac396bf926dc386b8fe3cebcd1b60a498a19c8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/onyx-dot-app/onyx/pull/12227?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: backend/onyx/server/features/mcp/api.py [hunk 1]
```diff
@@ -2536,6 +2587,16 @@ def create_mcp_server_simple(
...
+        mcp_server=mcp_server,
+        acting_user=user,
+        is_public=request.is_public,
+        user_ids=request.users,
+        group_ids=request.groups,
+        is_new=True,
+        db_session=db_session,
+    )
+
     db_session.commit()
```
[the remark this answers is not in the document]
Comment (beraterkanelcelik) [reply backend/onyx/server/features/mcp/api.py:2620]: This is handled: the `make_mcp_server_private` call in `_apply_mcp_server_access` is wrapped in `try/except NotImplementedError` and re-raised as `OnyxError(OnyxErrorCode.EE_REQUIRED)`, which the global exception handler turns into a clean 403 (not a 500). Since both the create and update simple endpoints (and the tool-config upsert) route through this one helper, all call sites are covered. Verified locally: restricting in CE returns 403 `EE_REQUIRED`.

**Comment by beraterkanelcelik** [comment]
@evan-onyx when you have a moment — this adds **group/public access control for MCP servers**, so admins can mark a server public or restrict it to specific user groups, and only users with access can add its tools to their agents (existing/unrestricted servers default to public, so it's backward-compatible).

It reuses the already-present `MCPServer__User` / `MCPServer__UserGroup` tables (the read path had a `TODO` and ignored them) and mirrors the existing `Persona` / `DocumentSet` `is_public` + groups + versioned-writer pattern. CE ships the model, migration, access-read query, API wiring, persona-upsert security gate (so the gate can't be bypassed via the API), and the `IsPublicGroupSelector` UI; the one EE piece (the group/user row writer) is left for your side — happy to send the short spec.

cubic is clean; greptile's remaining P1 is an already-addressed item (the `NotImplementedError` is caught and re-raised as `OnyxError(EE_REQUIRED)` → 403). The red CI checks are fork-infra only (ECR/Linear). Glad to split this into stacked PRs (migration / backend / frontend) if you'd prefer.

**Comment by beraterkanelcelik** [comment]
Addressed the access-control gaps:
- `GET /mcp/server/{id}/tools` now gates non-admin callers via `user_can_access_mcp_server` (was an IDOR + outbound-connect against any private server).
- The global `GET /tool` listing now filters MCP tools to servers the caller can access.
- `_add_mcp_server_access_filter` includes `owner` so a curator still sees their own private server.
- `upsert_mcp_server` re-raises `OnyxError` (e.g. `EE_REQUIRED` 403) instead of masking it as a 500.

@greptile-apps @cubic-dev-ai please re-review.

**Comment by evan-onyx** [comment]
#12972 has been merged!
